pkg/controller/server/device/managed.go

```diff
@@ -182,6 +189,36 @@ func (e *external) Create(ctx context.Context, mg resource.Managed) (managed.Ext
 
 	d.Status.SetConditions(runtimev1alpha1.Creating())
 
+	// TODO(displague) clean up this hack to demonstrate userdataRef
+	if d.Spec.ForProvider.UserDataRef != nil {
+		configMapKey := "cloud-init"
+		errGetUserDataRef := "cannot get required ConfigMap for UserDataRef"
+
+		ref := d.Spec.ForProvider.UserDataRef
+		cm := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      ref.Name,
+				Namespace: ref.Namespace,
+			},
+		}
+		nsn := types.NamespacedName{
+			Name:      cm.GetName(),
+			Namespace: cm.GetNamespace(),
+		}
+		if err := e.kube.Get(ctx, nsn, cm); err != nil {
+			if !ref.Optional {
+				return managed.ExternalCreation{}, errors.Wrap(err, errGetUserDataRef)
+			}
+		}
+		key := ref.Key
+		if key == "" {
+			// TODO(displague) use default key, or use first key in configmap?
+			key = configMapKey
+		}
+		userdata := cm.Data[key]
+		d.Spec.ForProvider.UserData = &userdata
```

Comment:
I don't think we should modify UserData in spec like this, this is acting like a late initialization.

Ultimately, either userdata or userdataRef should be provided.